tenantcostclient: collect tenant usage at replica batch level

andy-kimball opened the pull request:
**tenantcostclient: collect tenant usage at replica batch level**
Today's tenant usage collection framework does not take the replication factor
into account. This does not matter much right now, because the replication factor
is always 3 for Serverless. But eventually we will allow developers to change the
replication factor for tables, and will need to charge proportionally more in that
case.

In addition, today we processes batches at the highest level of DistSender, before
batches have been split by range. The upcoming Streamer component pre-splits the
batches by range and breaks them up in unpredictable ways, which scrambles request
unit calculations.

Related to the above problems, DistSender cannot accurately cost either read or
write batches until the replication factor is known in the case of write batches,
and until a response is received in the case of read batches. This calls into
question its current behavior of throttling based on the content of requests. Read
batches in particular will not throttle until outstanding debt has reached fairly
high levels, because the read request has a cost close to zero.

Finally, if an error occurs during message transmission, RUs are still charged
against the tenant. We'd prefer to not charge customers if there are internal
network or server problems that are causing errors, since those conditions are
generally not caused by them or under their control.

The changes in this commit takes the replication factor into accounts by waiting at
a deeper level in DistSender (i.e. in replicaSend rather than Send). In addition,
rather than waiting based on the content of a read/write batch if its cost is
greater than the number of tokens in the bucket, the new algorihm allows the token
bucket to go into debt. At that point, both batch requests *and* batch responses
will block until the debt is paid via the fill rate or new tokens from the server.

In order to smooth out request/response flow, CPU and Egress usage is amortized
across read/write operations so that operation latency is more consistent rather
than piling up on the first operations that arrive at the beginning of a new tick
period.

This commit also contains new RU usage model settings, based on a larger number of
training scenarios, and so is quite a bit more accurate for scenarios like IMPORT
and TPC-C.

**tenantcostmodel: update tenant cost model**
The current tenant cost model only takes batches into account. It ignores the
requests in each read and write batch (e.g. GetRequest, CPutRequest, etc). For
example, lookup joins typically send a small number of read batches with a large
number of requests in each batch, which resulted in the model greatly under-
costing queries containing them.

This commit updates the cost model to take requests into account, as well as
batches.

andy-kimball committed:
tenantcostmodel: update tenant cost model

The current tenant cost model only takes batches into account. It ignores the
requests in each read and write batch (e.g. GetRequest, CPutRequest, etc). For
example, lookup joins typically send a small number of read batches with a large
number of requests in each batch, which resulted in the model greatly under-
costing queries containing them.

This commit updates the cost model to take requests into account, as well as
batches.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81275)
<!-- Reviewable:end -->

andy-kimball commented:
Given that Radu's OOO, I need to find other reviewers. @irfansharif and @kvoli, I thought you might take a look, since you're on the KV team (where all these changes are) and because you've recently poked around in this code. @ajwerner, I'm including you b/c you took a look at Radu's original distributed rate limiting PRs.

RaduBerinde reviewed (commented):
I can review, it might just take me a couple of days. I have some initial comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-:-N2DGABy0dL31QBNfl16:b-gginyz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, and @kvoli)

___
*[`-- commits` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DJ_dqDaONRwuBES8l):*
[nit] Worth mentioning here that the way we achieve this is to only charge RUs on completion (rather than charge at least part of the RUs upfront, like today).
___
*[`-- commits` line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DFEju3rzuTQtbWo14):*
Did you actually observe this "piling on every tick" behavior? There is a smoothing mechanism for debt payback - it tries to carve out part of the refill rate to pay off the debt over 2 seconds (#71149). It seems like part of this commit is trying to do something similar above the token bucket.. Are you sure this is needed? And if yes, how is it intended to interact with the debt smoothing mechanism in the token bucket?
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DEAptB2gVR7oKztnP:-N2DEApu7dbsbD3rI5Dd:b-ouols4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a65eef85198dbd478b22011fa6d6f5088cb9a50/pkg/multitenant/tenantcostmodel/model.go#L22)):*
> ```Go
> type RU float64
> 
> // replicationFactor is the the number of replicas to which writes are sent, for
> ```

[nit] This is not used and is removed by the second commit
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DHZeYAKMQVLjBGV38:-N2DHZeYAKMQVLjBGV39:b3n31f9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02908f9cef4c857df8fcf105b73e4f8d3d682b6/pkg/multitenant/tenantcostmodel/settings.go#L31)):*
> ```Go
> 	readBatchCost = settings.RegisterFloatSetting(
> 		settings.TenantReadOnly,
> 		"tenant_cost_model.read_batch_cost",
> ```

Might want to add the old settings to the "retiredSettings" set.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/testdata/amortize` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DI8w36S7FvONNIXve:-N2DI8w36S7FvONNIXvf:b798mhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02908f9cef4c857df8fcf105b73e4f8d3d682b6/pkg/ccl/multitenantccl/tenantcostclient/testdata/amortize#L1)):*
> ```
> # Test token bucket that has gone into debt.
> ```

[nit] You probably did this already but make sure to run these under `stress` and `stressrace`.. I found it very hard to make these kinds of tests non-flaky


<!-- Sent from Reviewable.io -->

andy-kimball commented:
OK, thanks, there's no particular hurry on this, so take your time.





<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`-- commits` line 39 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DJ_dqDaONRwuBES8l):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Worth mentioning here that the way we achieve this is to only charge RUs on completion (rather than charge at least part of the RUs upfront, like today).
</blockquote></details>

Done.
___
*[`-- commits` line 50 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DFEju3rzuTQtbWo14):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Did you actually observe this "piling on every tick" behavior? There is a smoothing mechanism for debt payback - it tries to carve out part of the refill rate to pay off the debt over 2 seconds (#71149). It seems like part of this commit is trying to do something similar above the token bucket.. Are you sure this is needed? And if yes, how is it intended to interact with the debt smoothing mechanism in the token bucket?
</blockquote></details>

I moved the smoothing up to the limiter from the token bucket. I changed the algorithm because I wanted to evenly distribute CPU/Egress charges across both read and write operations, both in debt and low-RU situations. We should talk offline more about this, because there's a lot to discuss.
___
*[`pkg/multitenant/tenantcostmodel/model.go` line 22 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DEAptB2gVR7oKztnP:-N2PoZmY6l6q_OUFF-79:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/8a65eef85198dbd478b22011fa6d6f5088cb9a50/pkg/multitenant/tenantcostmodel/model.go#L22)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] This is not used and is removed by the second commit
</blockquote></details>

Removed.
___
*[`pkg/multitenant/tenantcostmodel/settings.go` line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DHZeYAKMQVLjBGV38:-N2Ppl885zZK1wFda6Hn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02908f9cef4c857df8fcf105b73e4f8d3d682b6/pkg/multitenant/tenantcostmodel/settings.go#L31)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Might want to add the old settings to the "retiredSettings" set.
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/testdata/amortize` line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2DI8w36S7FvONNIXve:-N2Q-rec0mKOrBNzmig7:b-d6kxxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d02908f9cef4c857df8fcf105b73e4f8d3d682b6/pkg/ccl/multitenantccl/tenantcostclient/testdata/amortize#L1)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] You probably did this already but make sure to run these under `stress` and `stressrace`.. I found it very hard to make these kinds of tests non-flaky
</blockquote></details>

Yes, I did do that. You're right that it's challenging, but I think you built a nice test framework that makes it possible to control timing.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuO6eBpAt8dwWdITY:-N2WuO6eBpAt8dwWdITZ:bh3xfqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L29)):*
> ```Go
> // is cleared. The tenant-side controller will always call Wait with needed
> // tokens = 0 for batch requests, since it does not yet know the cost at that
> // time. The cost becomes available once the batch response has been recieved,
> ```

[nit] received
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuVbF2ju1JBmjjsHY:-N2WuVbF2ju1JBmjjsHZ:b54v3qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L40)):*
> ```Go
> // flow freely until the next second, where they are stopped again.
> //
> // To ensure more even flow, the limiter introduces the concept of "extra RU".
> ```

"extra RU" sounds like the opposite of what it is (as in "we have some extra RUs that we can use"), why not "RU debt" or "borrowed RU" or "loaned RU"
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuoIn3XXEjjw4hoF7:-N2WuoIn3XXEjjw4hoF8:bdlz0ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L117)):*
> ```Go
> 	if needed > 0 {
> 		// Combine the needed RUs with some portion of any outstanding extraRU.
> 		needed += l.amortizeExtraRU()
> ```

It feels sketchy that this can behave very differently with `needed=0` and `needed=epsilon`. We are using 0 as a special value that the caller uses to get different semantics.

I suggest adding a new API that replaces `Wait(0)`; that API can document this difference. Perhaps `Prepare()` or `WaitForNewOperation`.
___
*[`pkg/ccl/multitenantccl/tenantcostserver/system_table.go` line 588 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WsZ8D17bs41xXX3J3:-N2WsZ8EEFAD0futbLwB:b-3z3tpm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostserver/system_table.go#L588)):*
> ```Go
> 		tenant.Consumption.KVRU,
> 		tenant.Consumption.ReadRequests,
> 		tenant.Consumption.ReadBatches,
> ```

[nit] this change belongs in the previous commit
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2161 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WtSygE8S2MlTDwCzj:-N2WtSygE8S2MlTDwCzk:bf2354o) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/kv/kvclient/kvcoord/dist_sender.go#L2161)):*
> ```Go
> 					reqInfo := tenantcostmodel.MakeRequestInfo(&ba, numReplicas)
> 					respInfo := tenantcostmodel.MakeResponseInfo(br, !reqInfo.IsWrite())
> 					if err = ds.kvInterceptor.OnResponseWait(ctx, reqInfo, respInfo); err != nil {
> ```

[nit] `err := `


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XIjoj462mTFpWyCRS:-N2XIjoj462mTFpWyCRT:b-wclwzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L200)):*
> ```Go
> 			opsPerTick = 1
> 		}
> 		l.mu.extraRUPerOp = l.mu.extraUsage / tenantcostmodel.RU(opsPerTick)
> ```

It doesn't feel ideal that we evenly divide this extra RUs per op when ops can have vastly different costs.

What do you think about keeping track of (a recent average of) total RU needed per tick, and make the extra RUs for each op be proportional to its needed RUs?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XOdKI1reTVcewuG6r:-N2XOdKI1reTVcewuG6s:b-jhoj56) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L132)):*
> ```Go
> 	})
> 	// Subtract the RUs for currently waiting requests.
> 	result -= tenantcostmodel.RU(atomic.LoadInt64(&l.waitingRU))
> ```

The waiting RU here was needed because the token bucket state doesn't reflect currently waiting requests. Say we only have 100 tokens and now we suddenly get a request for 10,000 tokens. That request will be waiting, but the 10,000 won't show up as debt in `tb.AvailableTokens()` (which will return 100). If we don't count the waiting RUs, we won't know that we need to ask for 10,000 extra tokens from the server (see `sendTokenBucketRequest`). I'm not sure but I believe that the `large-request` testcase was covering this aspect.

Ideally the underlying token bucket would count those somehow but it is difficult to implement with the Acquire/TryToFullfill "retry loop" design.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 214 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XMaBHCrqMozmA8b6V:-N2XMaBHCrqMozmA8b6W:b-do40po) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L214)):*
> ```Go
> 	l.mu.Lock()
> 	defer l.mu.Unlock()
> ```

It's unfortunate that we take yet another lock on the hot path.. We could potentially move this inside `waitRequest.Acquire`; we'd need to make sure we only do it the first time `Acquire` is called for that request.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XQXFW0ue0qJdZ3rNi:-N2XQXFW0ue0qJdZ3rNj:b-p9tfon) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request#L5)):*
> ```
> #    notification threshold.
> #  - the bucket has more RUs available than the notification threshold.
> ```

How come this test is removed? It's testing a pretty important aspect (if waiting requests don't trigger a notification, we could be waiting for a very long time even though the server would be able to give us more RUs).


<!-- Sent from Reviewable.io -->

andy-kimball committed:
tenantcostclient: collect tenant usage at replica batch level

Today's tenant usage collection framework does not take the replication factor
into account. This does not matter much right now, because the replication factor
is always 3 for Serverless. But eventually we will allow developers to change the
replication factor for tables, and will need to charge proportionally more in that
case.

In addition, today we process batches at the highest level of DistSender, before
batches have been split by range. The upcoming Streamer component pre-splits the
batches by range and breaks them up in unpredictable ways, which scrambles request
unit calculations.

Related to the above problems, DistSender cannot accurately cost either read or
write batches until the replication factor is known in the case of write batches,
and until a response is received in the case of read batches. This calls into
question its current behavior of throttling based on the content of requests. Read
batches in particular will not throttle until outstanding debt has reached fairly
high levels, because the read request has a cost close to zero.

Finally, if an error occurs during message transmission, RUs are still charged
against the tenant. We'd prefer to not charge customers if there are internal
network or server problems that are causing errors, since those conditions are
generally not caused by them or under their control.

The changes in this commit takes the replication factor into accounts by waiting at
a deeper level in DistSender (i.e. in replicaSend rather than Send). In addition,
both batch requests *and* batch responses will block if the token bucket does not
have sufficient tokens, until the debt is paid via the fill rate or new tokens from
the server. In order to avoid charging RUs for error cases, we now will only charge
RUs at the time of a successful response, not at the time of the request as before.

This commit also contains new RU usage model settings, based on a larger number of
training scenarios, and so is quite a bit more accurate for scenarios like IMPORT
and TPC-C.

Release note: None

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XOdKI1reTVcewuG6r:-N2XkRDmDUfbvwQexpCn:bh3m4mj) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The waiting RU here was needed because the token bucket state doesn't reflect currently waiting requests. Say we only have 100 tokens and now we suddenly get a request for 10,000 tokens. That request will be waiting, but the 10,000 won't show up as debt in `tb.AvailableTokens()` (which will return 100). If we don't count the waiting RUs, we won't know that we need to ask for 10,000 extra tokens from the server (see `sendTokenBucketRequest`). I'm not sure but I believe that the `large-request` testcase was covering this aspect.

Ideally the underlying token bucket would count those somehow but it is difficult to implement with the Acquire/TryToFullfill "retry loop" design.
</blockquote></details>

One problem is that the only thing we're counting here is the RUs needed by _the head of the waiting queue_. Any other wait requests that arrive will get enqueued, but never added to `waitingRU` (because the quota pool won't call `Acquire` when the queue length is > 0). I also worry about requests that have their context canceled, leading to a "leak" of `waitingRU`, since they'll never get fulfilled.

My imperfect "solution" to these problems was to allow the head of the queue to put the token bucket into debt. In other words, if there are only 100 tokens and the 10K request comes in, I'll allow it to be fulfilled. Now the token bucket is at -9900 tokens, and then further requests will be blocked. This will have the same behavior as the old code re: `AvailableTokens`, since that will reflect the debt. And yet it avoids the context cancelation leak and means slightly simpler code.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 29 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuO6eBpAt8dwWdITY:-N2YKcENDfDpKypEiVh6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L29)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] received
</blockquote></details>

Done.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 40 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuVbF2ju1JBmjjsHY:-N2XOy8n5xmghO3YeaBo:b-moxa31) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L40)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"extra RU" sounds like the opposite of what it is (as in "we have some extra RUs that we can use"), why not "RU debt" or "borrowed RU" or "loaned RU"
</blockquote></details>

Good point, I switched to debt RU.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2WuoIn3XXEjjw4hoF7:-N2Xjxsf6LQLPGuInE-E:bfh6s6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L117)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It feels sketchy that this can behave very differently with `needed=0` and `needed=epsilon`. We are using 0 as a special value that the caller uses to get different semantics.

I suggest adding a new API that replaces `Wait(0)`; that API can document this difference. Perhaps `Prepare()` or `WaitForNewOperation`.
</blockquote></details>

Agreed, I split this into `OnRequestWait` and `OnResponseWait`.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XIjoj462mTFpWyCRS:-N2Xp7No2e8JJ0KQB9Sz:b-i0p8bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L200)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It doesn't feel ideal that we evenly divide this extra RUs per op when ops can have vastly different costs.

What do you think about keeping track of (a recent average of) total RU needed per tick, and make the extra RUs for each op be proportional to its needed RUs?
</blockquote></details>

I did originally have that, and switched to the simpler ops/tick concept, simply b/c I didn't think it mattered much either way. I could switch to the proportional allocation if you think it's preferable.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 214 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XMaBHCrqMozmA8b6V:-N2YKT1v491qyaQUmzHl:b-fh3jz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L214)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's unfortunate that we take yet another lock on the hot path.. We could potentially move this inside `waitRequest.Acquire`; we'd need to make sure we only do it the first time `Acquire` is called for that request.
</blockquote></details>

So you're thinking that we'd mutate the `waitRequest` inside `Acquire`? Maybe have a separate `debtRU` field that we'd set the first time `Acquire` was called?

More generally, are you thinking we'd try to get rid of the extra mutex altogether and always sync on the quota pool's mutex, for all fields?
___
*[`pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XQXFW0ue0qJdZ3rNi:-N2YeWM-7O_zL404Y21W:b-s5iy7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request#L5)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

How come this test is removed? It's testing a pretty important aspect (if waiting requests don't trigger a notification, we could be waiting for a very long time even though the server would be able to give us more RUs).
</blockquote></details>

Because this PR changes the bucket so that it doesn't block until it's already in debt. That 20000000 byte write will be fulfilled and trigger the low RU notification. Multiple other tests are already ensuring that's true.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XOdKI1reTVcewuG6r:-N2YqpwR1uxBygrt_DRY:b88yol4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L132)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

One problem is that the only thing we're counting here is the RUs needed by _the head of the waiting queue_. Any other wait requests that arrive will get enqueued, but never added to `waitingRU` (because the quota pool won't call `Acquire` when the queue length is > 0). I also worry about requests that have their context canceled, leading to a "leak" of `waitingRU`, since they'll never get fulfilled.

My imperfect "solution" to these problems was to allow the head of the queue to put the token bucket into debt. In other words, if there are only 100 tokens and the 10K request comes in, I'll allow it to be fulfilled. Now the token bucket is at -9900 tokens, and then further requests will be blocked. This will have the same behavior as the old code re: `AvailableTokens`, since that will reflect the debt. And yet it avoids the context cancelation leak and means slightly simpler code.
</blockquote></details>

Are you sure? I thought that onWaitStart is called for all waiting requests, not just the head of the queue. And that it is always paired with an onWaitFinish call. That is what I think is the case when I loom at QuotaPool.Acquire (which we do call directly in Wait, unlike waitRequest.Acquire)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XOdKI1reTVcewuG6r:-N2ZGRrl-wQ4UMoDWMHQ:bhmhpf7) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are you sure? I thought that onWaitStart is called for all waiting requests, not just the head of the queue. And that it is always paired with an onWaitFinish call. That is what I think is the case when I loom at QuotaPool.Acquire (which we do call directly in Wait, unlike waitRequest.Acquire)
</blockquote></details>

I'm referring to this block in the onWaitStart callback, which does run for all waiting requests:
```
		// Account for the RUs, unless we already did in waitRequest.Acquire.
		// This is necessary because Acquire is only called for the head of the
		// queue.
		if !req.waitingRUAccounted {
			req.waitingRUAccounted = true
			atomic.AddInt64(&l.waitingRU, req.neededCeil())
			if l.testInstr != nil {
				l.testInstr.Event(l.timeSource.Now(), WaitingRUAccountedInCallback)
			}
		}
```

Although it does look like the corresponding code is missing from onWaitFinish (for the cancelation case).


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XIjoj462mTFpWyCRS:-N2ZIIRkE8S_c9GASWCy:bonmpeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L200)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I did originally have that, and switched to the simpler ops/tick concept, simply b/c I didn't think it mattered much either way. I could switch to the proportional allocation if you think it's preferable.
</blockquote></details>

I don't feel strongly, just a thought.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 214 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XMaBHCrqMozmA8b6V:-N2ZH-Kf34NMysiGgKW7:bh1mj7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L214)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

So you're thinking that we'd mutate the `waitRequest` inside `Acquire`? Maybe have a separate `debtRU` field that we'd set the first time `Acquire` was called?

More generally, are you thinking we'd try to get rid of the extra mutex altogether and always sync on the quota pool's mutex, for all fields?
</blockquote></details>

Yeah, exactly.

It would be great to use the quotapool mutex, but I didn't see a clean way of doing that. I did try a few refactorings of the token bucket - one in which you could just Lock the pool, and another where you would provide a reference to an existing mutex when creating the pool - but I abandoned both (I don't remember the specifics, just that a bunch of use cases got worse).


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @kvoli, and @RaduBerinde)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XOdKI1reTVcewuG6r:-N2b91ZaAMDh-_Nv6EWI:b-45lcmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L132)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm referring to this block in the onWaitStart callback, which does run for all waiting requests:
```
		// Account for the RUs, unless we already did in waitRequest.Acquire.
		// This is necessary because Acquire is only called for the head of the
		// queue.
		if !req.waitingRUAccounted {
			req.waitingRUAccounted = true
			atomic.AddInt64(&l.waitingRU, req.neededCeil())
			if l.testInstr != nil {
				l.testInstr.Event(l.timeSource.Now(), WaitingRUAccountedInCallback)
			}
		}
```

Although it does look like the corresponding code is missing from onWaitFinish (for the cancelation case).
</blockquote></details>

You're right, I missed the significance of that code and how `onWaitStart` works. I've restored the `waitingRU` concept. However, I decided to protect it with the quotapool mutex rather than atomic ops because my refactoring made that work pretty well.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 200 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XIjoj462mTFpWyCRS:-N37p7Ag-WfcuUsqPUBO:bz4pukv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L200)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't feel strongly, just a thought.
</blockquote></details>

I updated this to use a proportional allocation.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 214 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XMaBHCrqMozmA8b6V:-N37pHzj9nDTZRJW9izm:b-up9rv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7023a53a685b438cd0b39ee6f463c3ba6ab772c5/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L214)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah, exactly.

It would be great to use the quotapool mutex, but I didn't see a clean way of doing that. I did try a few refactorings of the token bucket - one in which you could just Lock the pool, and another where you would provide a reference to an existing mutex when creating the pool - but I abandoned both (I don't remember the specifics, just that a bunch of use cases got worse).
</blockquote></details>

I switched the limiter to always use the quotapool mutex. See if you think it looks clean enough / will perform well. It seems OK to me.
___
*[`pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N2XQXFW0ue0qJdZ3rNi:-N381c7IEg5DqlGTuZlh:b-yn2v15) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac5c05fd582c4530fec3cbb3ed4662325e46eb5d/pkg/ccl/multitenantccl/tenantcostclient/testdata/large-request#L5)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Because this PR changes the bucket so that it doesn't block until it's already in debt. That 20000000 byte write will be fulfilled and trigger the low RU notification. Multiple other tests are already ensuring that's true.
</blockquote></details>

Because I added waitingRU back, I restored this test and expanded it as well.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
I decided to go back to a variation of your original token bucket code. I had various problems with it originally, which caused me to implement it at the higher level. In particular, these lines in `TryToFulfill` were problematic:
```
	if amount <= tb.available {
		tb.available -= amount
		return true, 0
	}
```

The trouble was that we now pass `amount=0` for every read/write request. We want to block if the token bucket is too far in debt. But the old token bucket code was never blocking, because `amount <= tb.available` was always true.

The solution I came up with for this problem is to divide debt into two portions. Any debt which cannot be paid within two seconds is directly subtracted from `tb.available`. Any debt which can be paid within two seconds is stored in `tb.waitingDebt` and is smoothly applied to `tb.available` using your original algorithm. Allowing `tb.available` to go negative solves the problem, since now if there's so much debt that we can't process it quickly enough, all fulfillment requests will block, even if they request 0 tokens.





<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Unless I'm missing something, it looks like in the latest version we are no longer accounting the waiting RUs for any but the first waiting request. We would need to do that in onWaitStart.

RaduBerinde commented:
Well, maybe this is no big deal since now essentially all waiting requests have RU=0? But even so we should update the documentation for what waitingRUs is (or just remove it altogether).

andy-kimball reviewed (commented):
You're right that we need it in `onWaitStart`. There's still the case where there are many concurrent read/write responses with RU <> 0 that stack up. I'd added and removed `onWaitStart` several times, and just neglected to add it back in the final PR. So I put it back in and also added a test that ensures it's working.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-:-N3c0PJD81yfK2H-_qa2:bqt4bc7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @irfansharif, @kvoli, @miretskiy, @RaduBerinde, and @stevendanna)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

Sorry for the delay from my side (I was traveling and it was next to impossible to review on my phone).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-:-N4T0jOv6qbcgPDL-Mka:b-6ym83z)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, @miretskiy, @RaduBerinde, and @stevendanna)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 50 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4Sz6Cx6WtAoWGYQ9gZ:-N4Sz6Cx6WtAoWGYQ9g_:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L50)):*
> ```Go
> 	// them in the scope of the abstract pool's Update method or within the
> 	// wait request's Acquire method.
> 	qp struct {
> ```

Nice!
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 157 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4SzPebCMJraaLWO4wm:-N4SzPebCMJraaLWO4wn:b-slnixg) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L157)):*
> ```Go
> 	l.qp.Update(func(quotapool.Resource) (shouldNotify bool) {
> 		// If we already produced a notification that wasn't processed, drain it.
> 		// This not racy as long as this code does not run in parallel with the
> ```

[nit] What would be racy if the other receiver ran in parallel?
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 177 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4T-5m5CkpfMPvznZoF:-N4T-5m5CkpfMPvznZoG:bme7eee) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L177)):*
> ```Go
> }
> 
> // SetupNotification is used to call tokenBucket.SetupNotification under the
> ```

[nit] I don't think there is a `tokenBucket.SetupNotification` anymore
___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 210 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4T-JOqEo1Tt8ul4RWw:-N4T-JOqEo1Tt8ul4RWx:b-cblr27) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L210)):*
> ```Go
> }
> 
> // notify tries to send a non-blocking notification on lowRUNotify and disables
> ```

[nit] "tries" is a bit vague, maybe "sends a non-blocking notification on lowRUNotify (unless there is already a notification pending in the channel)"
___
*[`pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go` line 816 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4T0EG_Bn6whyciCNLv:-N4T0EG_Bn6whyciCNLw:b-alkkzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/tenant_side.go#L816)):*
> ```Go
> 
> // onExternalIO records external I/O usage, optionally waiting until there are
> // sufficient tokens in the bucket.
> ```

[nit] should explain in what cases we can get an error here. I think only if we're shutting down? (and then update the comment at the callsite above as well)


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
No problem, thanks for taking a look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-:-N4WYiWY8W2tYbba3ajo:b3kf4zt)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @irfansharif, @kvoli, @miretskiy, @RaduBerinde, and @stevendanna)

___
*[`pkg/ccl/multitenantccl/tenantcostclient/limiter.go` line 157 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/81275#-N4SzPebCMJraaLWO4wm:-N4WWeQD0OzLT56E2Jvp:b2asqdy) ([raw file](https://github.com/cockroachdb/cockroach/blob/0e01059c414cd5da9d2ec91bea0dbf30d26eb93a/pkg/ccl/multitenantccl/tenantcostclient/limiter.go#L157)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] What would be racy if the other receiver ran in parallel?
</blockquote></details>

This was your comment, I just copied it from the token bucket. : )

I don't see any problematic races here, so I'll just remove this comment.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+
